### PR TITLE
c8d/push: Support pushing all tags 

### PR DESCRIPTION
**Pull request by vvoland**
**c8d/push: Support pushing all tags**

Implement missing feature that pushes all tags from the provided local
repository.

**c8d/push: Add missing message about repository**

(a bit unrelated but I decided to include it here since it's only one line and I'm already making changes here)

Add "The push referers to repository X" message which is present in the
push output when using the graphdrivers.

**- How to verify it**
[c8d integration tests CI](https://github.com/moby/moby/actions/runs/6187498841/job/16797538706?pr=45232)
```diff
-DONE 277 tests, 5 skipped, 54 failures in 447.508s
+DONE 277 tests, 5 skipped, 47 failures in 501.680s
```

```bash
$ make TEST_FILTER='TestPushMultipleTags'  TEST_IGNORE_CGROUP_CHECK=1 DOCKERCLI_INTEGRATION_VERSION=v24.0.5  DOCKER_GRAPHDRIVER=overlayfs TEST_INTEGRATION_USE_SNAPSHOTTER=1 test-integration
...
=== RUN   TestDockerRegistrySuite/TestPushMultipleTags
--- PASS: TestDockerRegistrySuite (0.27s)
    --- PASS: TestDockerRegistrySuite/TestPushMultipleTags (0.27s)
```

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: daemon/containerd/image_push.go [hunk 1]
```diff
@@ -36,13 +39,46 @@ import (
...
+			// Image is not tagged nor digested, that means all tags push was requested.
+
+			// Find all images with the same repository.
+			nameFilter := "^" + regexp.QuoteMeta(sourceRef.Name()) + ":" + reference.TagRegexp.String() + "$"
+			imgs, err := i.client.ImageService().List(ctx, "name~="+strconv.Quote(nameFilter))
+			if err != nil {
+				return err
+			}
+
+			for _, img := range imgs {
```
Comment (rumpl) [inline daemon/containerd/image_push.go:57]: Does the other implementation also push tags one by one? If not maybe we should add a TODO here

Comment (vvoland) [reply daemon/containerd/image_push.go:57]: Yes: https://github.com/moby/moby/blob/20f96354699d20eeb4674d3944486b5d73c256b2/distribution/push_v2.go#L109-L118

**Review by rumpl — APPROVED** [review]

File: daemon/containerd/image_push.go [hunk 2]
```diff
@@ -36,13 +39,45 @@ import (
...
+				named, err := reference.ParseNamed(img.Name)
+				if err != nil {
+					// This shouldn't happen, but log a warning just in case.
+					log.G(ctx).WithFields(log.Fields{
+						"image":     img.Name,
+						"sourceRef": sourceRef,
+					}).Warn("refusing to push a invalid tag")
```
Comment (thaJeztah) [inline daemon/containerd/image_push.go:64]: minor grammar mistake;

```suggestion
					}).Warn("refusing to push an invalid tag")
```

File: daemon/containerd/image_push.go [hunk 3]
```diff
@@ -36,13 +39,45 @@ import (
...
+			return nil
 		}
 	}
 
+	return i.pushRef(ctx, sourceRef, metaHeaders, authConfig, out)
+}
+
+func (i *ImageService) pushRef(ctx context.Context, targetRef reference.Named, metaHeaders map[string][]string, authConfig *registry.AuthConfig, out progress.Output) (retErr error) {
+
 	leasedCtx, release, err := i.client.WithLease(ctx)
```
Comment (thaJeztah) [inline daemon/containerd/image_push.go:82]: If you decide to update, can you remove the empty line here?

**Review by thaJeztah — APPROVED** [review]
LGTM, but left two tiny comments; let me know if you want to update those, otherwise feel free to merge

File: daemon/containerd/image_push.go (hunk 2 above)
Comment (vvoland) [reply daemon/containerd/image_push.go:64]: Will update, thanks!

File: daemon/containerd/image_push.go (hunk 3 above)
Comment (vvoland) [reply daemon/containerd/image_push.go:82]: Sure!

**Pull request merged**